File: docs/code_samples/processing/c#/start_processing.cs
using System;
using System.Net.Http;
using System.Threading.Tasks;
using System.Text;
using System.Text.Json;
using System.IO;
using System.Collections.Generic;
using System.Linq;

var basePath = Directory.GetCurrentDirectory();
var baseUrl = "http://dev-api.v2.areal.ai/api/v2";

// 0. Login - details in Authentication section
// See code_samples/auth/c#/login.cs for authentication

var handler = new HttpClientHandler();
var client = new HttpClient(handler);
// Assume client is already authenticated with cookies from login

// 1. Get Pre-Signed URL's for a secure & fast upload channel
var fileNames = new[] { "sample.pdf" };
var presignedUrlRequest = new
{
    file_names = fileNames
};

var presignedUrlContent = new StringContent(
    JsonSerializer.Serialize(presignedUrlRequest),
    Encoding.UTF8,
    "application/json"
);

var presignedUrlResponse = await client.PostAsync($"{baseUrl}/processing/presigned_url/", presignedUrlContent);
presignedUrlResponse.EnsureSuccessStatusCode();

var presignedUrlResponseJson = await presignedUrlResponse.Content.ReadAsStringAsync();
var presignedUrlData = JsonSerializer.Deserialize<JsonElement>(presignedUrlResponseJson);

// 2. Upload to your PDF's to the PreSignedURL's
var presignedUrls = presignedUrlData.GetProperty("presigned_urls").EnumerateArray();
var uploadSessionId = presignedUrlData.GetProperty("upload_session_id").GetString();

foreach (var (fileName, presignedUrl) in fileNames.Zip(presignedUrls))
{
    var url = presignedUrl.GetProperty("url").GetString();
    var documentId = presignedUrl.GetProperty("document_id").GetString();
    var fields = presignedUrl.GetProperty("fields");
    
    using var uploadClient = new HttpClient();
    var formData = new MultipartFormDataContent();
    
    // Add fields from presigned URL
    foreach (var field in fields.EnumerateObject())
    {
        formData.Add(new StringContent(field.Value.GetString()), field.Name);
    }
    
    // Add file
    var fileBytes = await File.ReadAllBytesAsync(Path.Combine(basePath, fileName));
    formData.Add(new ByteArrayContent(fileBytes), "file", fileName);
    
    // 2. Upload to your PDF's to the PreSignedURL's
    var uploadResponse = await uploadClient.PostAsync(url, formData);
    uploadResponse.EnsureSuccessStatusCode();
    
    // 3. Start processing flow
    var processRequest = new
    {
        upload_session_id = uploadSessionId,
        document_id = documentId,
        file_name = fileName
    };
    
    var processContent = new StringContent(
        JsonSerializer.Serialize(processRequest),
        Encoding.UTF8,
        "application/json"
    );
    
    var processResponse = await client.PostAsync($"{baseUrl}/processing/upload/", processContent);
    processResponse.EnsureSuccessStatusCode();
}

File: docs/code_samples/processing/python/start_processing.py
import requests
from pathlib import Path

BASE_PATH = Path(__file__).parent
BASE_URL = 'http://dev-api.v2.areal.ai/api/v2'

# 0. Login - details in Authentication section
...

# 1. Get Pre-Signed URL's for a secure & fast upload channel
file_names = ['sample.pdf']
presigned_url_response = client.post(
    f'{BASE_URL}/processing/presigned_url/',
    # NOTE: you can upload to an existing session by passing the upload_session_id
    # but we recommend creating a new session for each loan
    # params={'upload_session_id': upload_session_id},
    json={'file_names': file_names},
).json()

# 2. Upload to your PDF's to the PreSignedURL's
for file_name, presigned_url in zip(
    file_names, presigned_url_response['presigned_urls']
):
    upload_response = requests.post(
        presigned_url['url'],
        data=presigned_url['fields'],
        files={'file': (BASE_PATH / file_name).read_bytes()},
    )

    # 3. Start processing flow
    process_response = client.post(
        f'{BASE_URL}/processing/upload/',
        json={
            'upload_session_id': presigned_url_response['upload_session_id'],
            'document_id': presigned_url['document_id'],
            'file_name': file_name,
        },
    )
